test_util/marathon_api.py

```diff
@@ -0,0 +1,242 @@
+import collections
+import copy
+import logging
+import uuid
+from contextlib import contextmanager
+
+import requests
+import retrying
+
+import test_util.helpers
+
+TEST_APP_NAME_FMT = '/integration-test-{}'
+
+
+def get_test_app(custom_port=False):
+    test_uuid = uuid.uuid4().hex
+    app = copy.deepcopy({
+        'id': TEST_APP_NAME_FMT.format(test_uuid),
+        'cpus': 0.1,
+        'mem': 32,
+        'instances': 1,
+        # NOTE: uses '.' rather than `source`, since `source` only exists in bash and this is
+        # run by sh
+        'cmd': '. /opt/mesosphere/environment.export && /opt/mesosphere/bin/python '
+               '/opt/mesosphere/active/dcos-integration-test/python_test_server.py ',
+        'env': {
+            'DCOS_TEST_UUID': test_uuid,
+            # required for python_test_server.py to run as nobody
+            'HOME': '/'
+        },
+        'healthChecks': [
+            {
+                'protocol': 'HTTP',
+                'path': '/ping',
+                'portIndex': 0,
+                'gracePeriodSeconds': 5,
+                'intervalSeconds': 10,
+                'timeoutSeconds': 10,
+                'maxConsecutiveFailures': 3
+            }
+        ],
+    })
+    if not custom_port:
+        app['cmd'] += '$PORT0'
+        app['portDefinitions'] = [{
+            "protocol": "tcp",
+            "port": 0,
+            "name": "test"
+        }]
+    return app, test_uuid
+
+
+def get_test_app_in_docker(ip_per_container):
+    app, test_uuid = get_test_app(custom_port=True)
+    assert 'portDefinitions' not in app
+    app['cmd'] += '9080'  # Fixed port for inside bridge networking or IP per container
+    app['container'] = {
+        'type': 'DOCKER',
+        'docker': {
+            # TODO(cmaloney): Switch to alpine with glibc
+            'image': 'debian:jessie',
+            'portMappings': [{
+                'hostPort': 0,
+                'containerPort': 9080,
+                'protocol': 'tcp',
+                'name': 'test',
+                'labels': {}
+            }]},
+        'volumes': [{
+            'containerPath': '/opt/mesosphere',
+            'hostPath': '/opt/mesosphere',
+            'mode': 'RO'
+        }]
+    }
+    if ip_per_container:
+        app['container']['docker']['network'] = 'USER'
+        app['ipAddress'] = {'networkName': 'dcos'}
+    else:
+        app['container']['docker']['network'] = 'BRIDGE'
+    return app, test_uuid
+
+
+class MarathonApi(test_util.helpers.ApiClient):
+    @classmethod
+    def get_client(cls, cluster, user):
```

Comment:
why can't this be in the constructor? Wrap the user provided default_headers with one new header?